lib/SILOptimizer/Mandatory/DiagnoseStaticExclusivity.cpp

```diff
@@ -0,0 +1,349 @@
+    NewRequires = SharedAccess;
+  } else {
+    NewRequires = ExclusiveAccess;
+  }
+
+  ExclusiveOrShared_t PriorRequires;
+  if (PriorAccess->getAccessKind() == SILAccessKind::Read) {
+    PriorRequires = SharedAccess;
+  } else {
+    PriorRequires = ExclusiveAccess;
+  }
+
+  diagnose(Ctx, NewAccess->getLoc().getSourceLoc(),
+           diag::exclusivity_access_required,
+           static_cast<unsigned>(NewAccess->getAccessKind()),
+           static_cast<unsigned>(NewRequires));
+  diagnose(Ctx, PriorAccess->getLoc().getSourceLoc(),
+           diag::exclusivity_conflicting_access,
+           static_cast<unsigned>(PriorAccess->getAccessKind()),
+           static_cast<unsigned>(PriorRequires));
+}
+
+/// Look through a value to find the underlying storage
+/// accessed.
+static AccessedStorage findAccessedStorage(SILValue Source) {
+  SILValue Iter = Source;
+  while (true) {
+    if (auto *PBI = dyn_cast<ProjectBoxInst>(Iter)) {
+      Iter = PBI->getOperand();
+    } else if (auto *CVI = dyn_cast<CopyValueInst>(Iter)) {
+      Iter = CVI->getOperand();
+    } else if (auto *GAI = dyn_cast<GlobalAddrInst>(Iter)) {
+      return AccessedStorage(GAI->getReferencedGlobal());
+    } else {
+      assert(Iter->getType().isAddress() || Iter->getType().is<SILBoxType>());
+      return AccessedStorage(Iter);
+    }
+  }
+}
+
+using StorageMap = llvm::SmallDenseMap<AccessedStorage, AccessInfo, 4>;
+
+static void checkStaticExclusivity(SILFunction &Fn) {
+  // The implementation relies on the following SIL invariants:
+  //    - All incoming edges to a block must have the same in-progress
+  //      accesses. This enables the analysis to not perform a data flow merge
+  //      on incoming edges.
+  //    - Further, for a given address each of the in-progress
+  //      accesses must have begun in the same order on all edges. This ensures
+  //      consistent diagnostics across changes to the exploration of the CFG.
+  //    - On return from a function there are no in-progress accesses. This
+  //      enables a sanity check for lean analysis state at function exit.
+  //    - Each end_access instruction corresponds to exactly one begin access
+  //      instruction. (This is encoded in the EndAccessInst itself)
+  //    - begin_access arguments cannot be basic block arguments.
+  //      This enables the analysis to look back to find the *single* storage
+  //      storage location accessed.
+
+  if (Fn.empty())
+    return;
+
+  // This is a staging flag. Eventually the ability to turn off static
+  // enforcement will be removed.
+  if (!Fn.getModule().getOptions().EnforceExclusivityStatic)
+    return;
+
+  // For each basic block, track the stack of current accesses on
+  // exit from that block.
+  llvm::SmallDenseMap<SILBasicBlock *, Optional<StorageMap>, 32>
+      BlockOutAccesses;
+
+  BlockOutAccesses[Fn.getEntryBlock()] = StorageMap();
+
+  llvm::ReversePostOrderTraversal<SILFunction *> RPOT(&Fn);
```

Comment:
One comment. You shouldn't use this. We have an analysis for computing a reverse post order. You should use that as well. This will ensure that if someone creates an RPO before you or after you reuse of the RPO can happen.